Data/HashMap/Base.hs

```diff
@@ -293,8 +297,11 @@ equal1 eq = go
       = bm1 == bm2 && A.sameArray1 go ary1 ary2
     go (Leaf h1 l1) (Leaf h2 l2) = h1 == h2 && leafEq l1 l2
     go (Full ary1) (Full ary2) = A.sameArray1 go ary1 ary2
-    go (Collision h1 ary1) (Collision h2 ary2)
-      = h1 == h2 && isPermutationBy leafEq (A.toList ary1) (A.toList ary2)
+    go (Collision h1 (L k11 v11) (L k12 v12) hm1) (Collision h2 (L k21 v21) (L k22 v22) hm2)
+      = h1 == h2
+      && isPermutationBy (\(k1, v1) (k2, v2) -> k1 == k2 && eq v1 v2)
+           ((k11, v11) : (k12, v12) : toList hm1)
+           ((k21, v21) : (k22, v22) : toList hm2)
```

Comment:
This is not (significantly) worse than what we have now, but it's still O(n^2) when collision buckets are very full. If we have a pathological chain of `Collision` nodes, then we have no hope. But of course that only happens when there are lots of collisions *and* salting doesn't help, which is an exceptionally rare situation. If we add a `Hashable k` constraint here (and therefore to the `Eq` and `Eq1` instances), we should be able to cut it down to linearithmic in the common case where salt works.

```haskell
go t1@(Collision h1 _ _ _) t2@(Collision h2 _ _ _) = h1 == h2 &&
   size t1 == size t2 && foldrWithKey (\k v r -> maybe False (eq v) (lookup k t2) && r) True t1
```